File: src/index/hotIndex.ts
import type { FileId, NameId, SymbolFlags, SymbolKind } from './types';

export interface HotEntry {
  fileId: FileId;
  kind: SymbolKind;
  flags: SymbolFlags;
  parentNameId?: NameId;
}

export class HotIndex {
  private byName = new Map<NameId, HotEntry[]>();
  private byFile = new Map<FileId, NameId[]>();

  add(nameId: NameId, entry: HotEntry): void {
    let entries = this.byName.get(nameId);
    if (!entries) {
      entries = [];
      this.byName.set(nameId, entries);
    }
    entries.push(entry);

    let names = this.byFile.get(entry.fileId);
    if (!names) {
      names = [];
      this.byFile.set(entry.fileId, names);
    }
    names.push(nameId);
  }

  removeFile(fileId: FileId): NameId[] {
    const names = this.byFile.get(fileId);
    if (!names) return [];
    for (const nameId of names) {
      const entries = this.byName.get(nameId);
      if (!entries) continue;
      const filtered = entries.filter((e) => e.fileId !== fileId);
      if (filtered.length === 0) {
        this.byName.delete(nameId);
      } else {
        this.byName.set(nameId, filtered);
      }
    }
    this.byFile.delete(fileId);
    return names;
  }

  lookup(nameId: NameId): readonly HotEntry[] | undefined {
    return this.byName.get(nameId);
  }

  hasName(nameId: NameId): boolean {
    return this.byName.has(nameId);
  }

  totalEntries(): number {
    let total = 0;
    for (const list of this.byName.values()) total += list.length;
    return total;
  }

  totalNames(): number {
    return this.byName.size;
  }
}

File: daemon/src/index/hot_index.rs
use std::collections::HashMap;

use super::string_table::StringId;
use super::symbol::SymbolKind;

pub type FileId = StringId;
pub type NameId = StringId;

#[derive(Debug, Clone)]
pub struct HotEntry {
    pub file_id: FileId,
    pub kind: SymbolKind,
    pub flags: u32,
    pub parent_name_id: Option<NameId>,
}

#[derive(Default)]
pub struct HotIndex {
    by_name: HashMap<NameId, Vec<HotEntry>>,
    by_file: HashMap<FileId, Vec<NameId>>,
}

impl HotIndex {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add(&mut self, name_id: NameId, entry: HotEntry) {
        self.by_file.entry(entry.file_id).or_default().push(name_id);
        self.by_name.entry(name_id).or_default().push(entry);
    }

    pub fn remove_file(&mut self, file_id: FileId) {
        let Some(names) = self.by_file.remove(&file_id) else {
            return;
        };
        for name_id in names {
            let Some(entries) = self.by_name.get_mut(&name_id) else {
                continue;
            };
            entries.retain(|e| e.file_id != file_id);
            if entries.is_empty() {
                self.by_name.remove(&name_id);
            }
        }
    }

    pub fn lookup(&self, name_id: NameId) -> Option<&[HotEntry]> {
        self.by_name.get(&name_id).map(|v| v.as_slice())
    }

    pub fn total_entries(&self) -> usize {
        self.by_name.values().map(|v| v.len()).sum()
    }
}
